jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PushImageStep.java

```diff
@@ -47,23 +45,25 @@
       ProgressEventDispatcher.Factory progressEventDispatcherFactory,
       RegistryClient registryClient,
       BlobDescriptor containerConfigurationDigestAndSize,
-      Image builtImage)
-      throws IOException {
+      BuildableManifestTemplate manifestTemplate,
+      DescriptorDigest imageDigest)
+      throws IOException, RegistryException {
     Set<String> tags = buildContext.getAllTargetImageTags();
 
+    EventHandlers eventHandlers = buildContext.getEventHandlers();
+
     try (TimerEventDispatcher ignored =
-            new TimerEventDispatcher(
-                buildContext.getEventHandlers(), "Preparing manifest pushers");
+            new TimerEventDispatcher(eventHandlers, "Preparing manifest pushers");
         ProgressEventDispatcher progressEventDispatcher =
             progressEventDispatcherFactory.create("launching manifest pushers", tags.size())) {
 
-      // Gets the image manifest to push.
-      BuildableManifestTemplate manifestTemplate =
-          new ImageToJsonTranslator(builtImage)
-              .getManifestTemplate(
-                  buildContext.getTargetFormat(), containerConfigurationDigestAndSize);
+      if (JibSystemProperties.skipExistingImages()
+          && registryClient.checkImage(imageDigest.toString()).isPresent()) {
```

Comment:
This check is still suffering from the concurrency issue I mentioned in https://github.com/GoogleContainerTools/jib/pull/2387#issuecomment-612192092. As we discussed, you should follow the [design outlined in the proposal]( https://github.com/GoogleContainerTools/jib/pull/2387/files#diff-7a6f7ffce8c0482a6cb7542dd553fd25R24-R28) to create a separate `-Step` class to call `registryClient.checkImage()`. If the program control already reached here, it means multiple `PushImageStep`s are scheduled and running.

Reply:
The `PushImageStep` steps are being created after the condition is evaluated to false. If it evaluates to true, no steps are created and an empty list is returned. It seems like it achieves the same behaviour but with simpler code. If you think that adding a new step is better, Ill change it.

Reply:
Oh, I jumped into the conclusion too early after I saw this check in this file.

However, you still need to create a separate Step for a blocking network call. `makeList()` should just create and return necessary Steps instantaneously and should never block thread scheduling in `StepsRunner`.

Reply:
Ok. Can you have a look at commit [987c58c](https://github.com/GoogleContainerTools/jib/pull/2531/commits/987c58cfec454be74a4c14c387c6e3446d43c47b)? I had initially introduced a step before the `makeList` call. You should look at `CheckImageStep`, `PushImageStep` and `StepsRunner`. Should I change it to something like that?

Reply:
The direction seems right. However, we maintain complete separation between business logic (what Steps do) and the async handling (what `StepsRunner` do). That is, `StepsRunner` doesn't care what each Step does or implies but is only interested in scheduling Steps, passing parameters to them as necessary, and ultimately storing results in `results.*`. All the business logic ([building a manifest and computing its digest](https://github.com/GoogleContainerTools/jib/pull/2531/commits/987c58cfec454be74a4c14c387c6e3446d43c47b#diff-18df945d9e250304dbf7dd948486e16fR395-R401) and [the conditional skip](https://github.com/GoogleContainerTools/jib/pull/2531/commits/987c58cfec454be74a4c14c387c6e3446d43c47b#diff-18df945d9e250304dbf7dd948486e16fR410-R414)) should go out of `StepsRunner`. For example, `StepsRunner` code shouldn't have conditional Step scheduling based on the result of some step. So, probably you should put the manifest construction and digest computation into `CheckImageStep`. The step would then return an `Optional<Manifest...>`, and `StepsRunner` would just store the result. (If deemed necessary, `CheckImageStep` could return the manifest it built, but because anyone can easily build the same manifest later, this may be unnecessary.) This manifest check result should probably be passed into `makeList()` where it returns either an empty list or multiple `PushImageStep` instances based on the result. The responsibility of `makeList()` should just be about properly instantiating necessary `-Step` instances where all the core business logic goes in.

Reply:
Pushed new changes for this.